dm/pkg/exstorage/utils.go

```diff
@@ -0,0 +1,148 @@
+// Copyright 2022 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package exstorage
+
+import (
+	"context"
+	"net/url"
+	"os"
+	"path"
+	"strings"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/tidb/br/pkg/storage"
+
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/service/s3"
+)
+
+// AdjustPath adjust rawURL, add uniqueId as path suffix.
+func AdjustPath(rawURL string, uniqueID string) (string, error) {
+	if rawURL == "" {
+		return rawURL, nil
+	}
+	u, err := url.Parse(rawURL)
+	if err != nil {
+		return "", errors.Trace(err)
+	}
+	trimPath := strings.TrimRight(u.Path, "/")
+	// avoid duplicate add uniqueID
+	if uniqueID != "" && !strings.HasSuffix(trimPath, uniqueID) {
+		u.Path = trimPath + "." + uniqueID
+		return u.String(), nil
+	}
+
+	return rawURL, nil
+}
+
+// isS3Path judges if rawURL is s3 path.
+func IsS3Path(rawURL string) (bool, error) {
+	if rawURL == "" {
+		return false, nil
+	}
+	u, err := url.Parse(rawURL)
+	if err != nil {
+		return false, err
+	}
+	if u.Scheme == "s3" {
+		return true, nil
+	}
+	return false, nil
+}
+
+// CreateExternalStore creates ExternalStore.
+func CreateExternalStore(ctx context.Context, path string) (storage.ExternalStorage, error) {
+	backend, err := storage.ParseBackend(path, nil)
+	if err != nil {
+		return nil, err
+	}
+	return storage.New(ctx, backend, &storage.ExternalStorageOptions{})
+}
+
+// CollectDirFiles gets files in dir.
+func CollectDirFiles(ctx context.Context, dir string, externalStore storage.ExternalStorage) (map[string]struct{}, error) {
```

Comment:
There adds a parameter `externalStore`, it is used to add unit test now, because it can not mock s3 storage create from a dir, and i used the way like https://github.com/pingcap/tidb/blob/master/br/pkg/storage/s3_test.go#L37.